File: python/valuecell/server/api/schemas/stock_analysis_research_feedback.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class StockAnalysisResearchFeedbackItemData(BaseModel):
    feedback_id: int
    thread_id: int
    user_id: str
    anchor_message_id: str
    title: str
    anchor_question_intent: str | None = None
    anchor_response_strategy: str | None = None
    anchor_mode: str | None = None
    anchor_plan_summary: str | None = None
    anchor_validation_status: str | None = None
    linked_task_ids_json: list[int] = Field(default_factory=list)
    linked_context_ids_json: list[int] = Field(default_factory=list)
    linked_memory_id: int | None = None
    linked_compression_id: int | None = None
    linked_compare_targets_json: list[dict[str, Any]] = Field(default_factory=list)
    linked_tickers_json: list[str] = Field(default_factory=list)
    linked_themes_json: list[str] = Field(default_factory=list)
    linked_outcome_review_ids_json: list[int] = Field(default_factory=list)
    linked_effectiveness_snapshot_json: dict[str, Any] = Field(default_factory=dict)
    linked_risk_sizing_snapshot_json: dict[str, Any] = Field(default_factory=dict)
    outcome_alignment_status: str
    process_quality_status: str
    compare_helpful: bool = False
    refresh_helpful: bool = False
    tooling_helpful: bool = False
    validation_helpful: bool = False
    what_helped_json: list[str] = Field(default_factory=list)
    what_hurt_json: list[str] = Field(default_factory=list)
    process_adjustments_json: list[str] = Field(default_factory=list)
    task_followup_suggestions_json: list[dict[str, Any]] = Field(default_factory=list)
    summary: str
    detail_note: str | None = None
    version: int = 1
    created_at: datetime
    updated_at: datetime


class StockAnalysisResearchFeedbackListData(BaseModel):
    thread_id: int
    items: list[StockAnalysisResearchFeedbackItemData] = Field(default_factory=list)
    latest_feedback: StockAnalysisResearchFeedbackItemData | None = None
    count: int = 0
    generated_at: datetime


class StockAnalysisResearchFeedbackCaptureRequest(BaseModel):
    anchor_message_id: str | None = None
    related_task_ids: list[int] = Field(default_factory=list)
    related_tickers: list[str] = Field(default_factory=list)
    title: str | None = None
    note: str | None = None


class StockAnalysisResearchFeedbackRefreshRequest(BaseModel):
    title: str | None = None
    note: str | None = None
    related_task_ids: list[int] = Field(default_factory=list)
    related_tickers: list[str] = Field(default_factory=list)


class StockAnalysisResearchFeedbackMutationData(BaseModel):
    thread_id: int
    feedback: StockAnalysisResearchFeedbackItemData

File: frontend/src/types/stock-analysis-research-feedback.ts
export type StockAnalysisResearchFeedbackTaskSuggestion = {
  task_id: number;
  title: string;
  current_status: string;
  suggestion: string;
  reason: string;
};

export type StockAnalysisResearchFeedback = {
  feedback_id: number;
  thread_id: number;
  user_id: string;
  anchor_message_id: string;
  title: string;
  anchor_question_intent?: string | null;
  anchor_response_strategy?: string | null;
  anchor_mode?: string | null;
  anchor_plan_summary?: string | null;
  anchor_validation_status?: string | null;
  linked_task_ids_json: number[];
  linked_context_ids_json: number[];
  linked_memory_id?: number | null;
  linked_compression_id?: number | null;
  linked_compare_targets_json: Array<Record<string, unknown>>;
  linked_tickers_json: string[];
  linked_themes_json: string[];
  linked_outcome_review_ids_json: number[];
  linked_effectiveness_snapshot_json: Record<string, unknown>;
  linked_risk_sizing_snapshot_json: Record<string, unknown>;
  outcome_alignment_status: string;
  process_quality_status: string;
  compare_helpful: boolean;
  refresh_helpful: boolean;
  tooling_helpful: boolean;
  validation_helpful: boolean;
  what_helped_json: string[];
  what_hurt_json: string[];
  process_adjustments_json: string[];
  task_followup_suggestions_json: StockAnalysisResearchFeedbackTaskSuggestion[];
  summary: string;
  detail_note?: string | null;
  version: number;
  created_at: string;
  updated_at: string;
};

export type StockAnalysisResearchFeedbackList = {
  thread_id: number;
  items: StockAnalysisResearchFeedback[];
  latest_feedback: StockAnalysisResearchFeedback | null;
  count: number;
  generated_at: string;
};

export type StockAnalysisResearchFeedbackMutationResult = {
  thread_id: number;
  feedback: StockAnalysisResearchFeedback;
};
